### PR TITLE
storage: Allow resizing of managed Stratis filesystems

### PR DESCRIPTION
**Pull request by mvollmer**
This is possible by setting their size limit, new in "r6".

Demo: https://www.youtube.com/watch?v=K4vRG8-eA4c

- [ ] Tests, coverage

### CODE REVIEW

File: pkg/storaged/client.js [hunk 1]
```diff
@@ -744,15 +744,18 @@ function update_indices() {
         const fsys_offsets = [];
         let fsys_total_used = 0;
         let fsys_total_size = 0;
+        let fsys_total_limit = 0;
         filesystems.forEach(fs => {
             fsys_offsets.push(fsys_total_used);
             fsys_total_used += fs.Used[0] ? Number(fs.Used[1]) : 0;
             fsys_total_size += Number(fs.Size);
+            fsys_total_limit += (fs.SizeLimit && fs.SizeLimit[0]) ? Number(fs.SizeLimit[1]) : Number(fs.Size);
```
Comment (cockpituous) [inline pkg/storaged/client.js:752]: This added line is not executed by any test.

File: pkg/storaged/client.js [hunk 2]
```diff
@@ -1398,12 +1403,16 @@
                             .input(passphrase);
                 };
 
+                client.stratis_set_property = (proxy, prop, sig, value) => {
```
Comment (cockpituous) [inline pkg/storaged/client.js:1406]: This added line is not executed by any test.

File: pkg/storaged/client.js [hunk 3]
```diff
@@ -1398,12 +1403,16 @@
...
+                    // DBusProxy is smart enough to allow "proxy.Prop
+                    // = value" to just work, but we want to catch any
+                    // error ourselves, and we want to wait for the
+                    // method call to complete.
+                    return stratis.call(proxy.path, "org.freedesktop.DBus.Properties", "Set",
+                                        [proxy.iface, prop, cockpit.variant(sig, value)]);
```
Comment (cockpituous) [inline pkg/storaged/client.js:1412]: These 2 added lines are not executed by any test.

File: pkg/storaged/stratis/create-dialog.jsx [hunk 4]
```diff
@@ -44,6 +44,12 @@ export function create_stratis_pool() {
             break;
     }
 
+    let managed_help;
+    if (client.features.stratis_size_limits)
+        managed_help = _("When this option is checked, the new pool will not allow overprovisioning. You need to specify a maximum size for each filesystem that is created in the pool. Filesystems can be made larger after creation but they can not always be made smaller. Snapshots are fully allocated on creation. The sum of all maximum sizes can not exceed the size of the pool. The advantage of this is that filesystems in this pool can not run out of space in a surprising way. The disadvantage is that you need to know the maximum size for each filesystem in advance and creation of snapshots is limited.");
```
Comment (cockpituous) [inline pkg/storaged/stratis/create-dialog.jsx:49]: This added line is not executed by any test.

File: pkg/storaged/stratis/create-dialog.jsx [hunk 5]
```diff
@@ -44,6 +44,12 @@
...
+    else
+        managed_help = _("When this option is checked, the new pool will not allow overprovisioning. You need to specify a maximum size for each filesystem that is created in the pool. Filesystems can not be resized after creation. Snapshots are fully allocated on creation. The sum of all maximum sizes can not exceed the size of the pool. The advantage of this is that filesystems in this pool can not run out of space in a surprising way. The disadvantage is that you need to know the maximum size for each filesystem in advance and creation of snapshots is limited.");
```
Comment (cockpituous) [inline pkg/storaged/stratis/create-dialog.jsx:51]: This added line is not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx [hunk 6]
```diff
@@ -206,6 +214,28 @@ const StratisFilesystemCard = ({
         });
     }
 
+    function set_limit() {
+        const size = fsys_size(pool, fsys);
+        dialog_open({
+            Title: cockpit.format(_("Resize filesystem $0"), fsys.Name),
+            Fields: [
+                SizeSlider("limit", _("Size"),
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:222]: These 6 added lines are not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx [hunk 7]
```diff
@@ -206,6 +214,28 @@
...
+                           {
+                               value: size,
+                               min: Number(fsys.Size),
+                               max: Number(size) + stats.pool_free,
+                               round: 512,
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:227]: These 4 added lines are not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx [hunk 8]
```diff
@@ -206,6 +214,28 @@
...
+                           }),
+            ],
+            Action: {
+                Title: _("Resize"),
+                action: async function (vals) {
+                    await client.stratis_set_property(fsys, "SizeLimit", "(bs)", [true, vals.limit.toString()]);
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:233]: These 4 added lines are not executed by any test.

File: pkg/storaged/stratis/pool.jsx [hunk 9]
```diff
@@ -115,15 +115,21 @@ function create_fs(pool) {
...
+                let result;
+                let size_spec = [false, ""];
+
+                if (managed_fsys_sizes)
+                    size_spec = [true, vals.size.toString()];
+
+                if (client.features.stratis_size_limits) {
+                    result = await pool.CreateFilesystems([[vals.name, size_spec, [false, ""]]]).then(std_reply);
+                } else {
+                    result = await pool.CreateFilesystems([[vals.name, size_spec]]).then(std_reply);
```
Comment (cockpituous) [inline pkg/storaged/stratis/pool.jsx:128]: This added line is not executed by any test.

File: pkg/storaged/stratis/pool.jsx [hunk 10]
```diff
@@ -502,7 +508,24 @@
         });
     }
 
-    const use = pool.TotalPhysicalUsed[0] && [Number(pool.TotalPhysicalUsed[1]), Number(pool.TotalPhysicalSize)];
+    let usage = null;
+    if (managed_fsys_sizes) {
+        usage = (
+            <StorageDescription title={_("Allocation")}>
+                <StorageUsageBar stats={[
+                                     stats.sensible_limits ? stats.fsys_total_limit : stats.fsys_total_size,
```
Comment (cockpituous) [inline pkg/storaged/stratis/pool.jsx:516]: This added line is not executed by any test.

File: pkg/storaged/stratis/utils.jsx [hunk 11]
```diff
@@ -182,3 +182,12 @@ export function set_mount_options(path, vals, forced_options) {
                         });
             });
 }
+
+export function fsys_size(pool, fsys) {
+    const stats = client.stratis_pool_stats[pool.path];
+
+    if (client.features.stratis_size_limits && stats.sensible_limits && fsys.SizeLimit[0])
+        return Number(fsys.SizeLimit[1]);
```
Comment (cockpituous) [inline pkg/storaged/stratis/utils.jsx:190]: This added line is not executed by any test.

**Comment by mvollmer** [comment]
oh, nice, r6 seems to be everywhere already. Hurray for procrastination!

**Comment by mvollmer** [comment]
This could maybe extended with a "Allow temporary overprovisioning" mode for a managed pool. This would set the overprov flag and put a mild warning on the screen that informs people that they need to be on the lookout of running out of space.

This might be part of creating snapshots: During creation of a snapshot, the pool can be put into temporary overprov, and after that no new snapshots and no new filesystems can be created. Deleting the snapshot will put the pool back into no-overprov mode. The mild warning might contain a action that deletes the snapshot as well.

A managed pool would then be recognized by either overprov being off, or by there being filesystems with limits.

File: pkg/storaged/client.js (hunk 1 above)
Comment (cockpituous) [inline pkg/storaged/client.js:752]: This added line is not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:222]: These 6 added lines are not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx (hunk 7 above)
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:227]: These 4 added lines are not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx (hunk 8 above)
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:233]: These 4 added lines are not executed by any test.

File: pkg/storaged/stratis/pool.jsx [hunk 12]
```diff
@@ -502,7 +503,24 @@ const StratisPoolCard = ({ card, pool, degraded_ops, can_grow, managed_fsys_size
         });
     }
 
-    const use = pool.TotalPhysicalUsed[0] && [Number(pool.TotalPhysicalUsed[1]), Number(pool.TotalPhysicalSize)];
+    let usage = null;
+    if (managed_fsys_sizes) {
+        usage = (
+            <StorageDescription title={_("Allocation")}>
+                <StorageUsageBar stats={[
+                    stats.sensible_limits ? stats.fsys_total_limit : stats.fsys_total_size,
```
Comment (cockpituous) [inline pkg/storaged/stratis/pool.jsx:511]: This added line is not executed by any test.

File: pkg/storaged/stratis/utils.jsx [hunk 13]
```diff
@@ -182,3 +182,12 @@ export function set_mount_options(path, vals, forced_options) {
                         });
             });
 }
+
+export function fsys_size(pool, fsys) {
+    const stats = client.stratis_pool_stats[pool.path];
+
+    if (stats.sensible_limits && fsys.SizeLimit[0])
+        return Number(fsys.SizeLimit[1]);
```
Comment (cockpituous) [inline pkg/storaged/stratis/utils.jsx:190]: This added line is not executed by any test.

**Comment by mvollmer** [comment]
Nope, change of plans.

**Pull request closed without merging**